File: src/test/java/com/gisiona/demodevinia/integration/UserIntegrationTest.java
package com.gisiona.demodevinia.integration;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.gisiona.demodevinia.infrastructure.adapter.web.dto.CreateUserRequest;
import com.gisiona.demodevinia.infrastructure.adapter.web.dto.UpdateUserRequest;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@AutoConfigureMockMvc
@TestPropertySource(properties = {
    "spring.datasource.url=jdbc:h2:mem:testdb-integration",
    "spring.jpa.hibernate.ddl-auto=create-drop"
})
@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
class UserIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Test
    void shouldCreateAndRetrieveUserSuccessfully() throws Exception {
        CreateUserRequest createRequest = new CreateUserRequest("João Silva", "joao@example.com");

        MvcResult createResult = mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.name").value("João Silva"))
                .andExpect(jsonPath("$.email").value("joao@example.com"))
                .andExpect(jsonPath("$.id").exists())
                .andReturn();

        String responseContent = createResult.getResponse().getContentAsString();
        Long userId = objectMapper.readTree(responseContent).get("id").asLong();

        mockMvc.perform(get("/api/users/{id}", userId))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(userId))
                .andExpect(jsonPath("$.name").value("João Silva"))
                .andExpect(jsonPath("$.email").value("joao@example.com"));
    }

    @Test
    void shouldNotCreateUserWithDuplicateEmail() throws Exception {
        CreateUserRequest createRequest = new CreateUserRequest("João Silva", "joao@example.com");

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isCreated());

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.message").value("Email já está em uso: joao@example.com"));
    }

    @Test
    void shouldUpdateUserSuccessfully() throws Exception {
        CreateUserRequest createRequest = new CreateUserRequest("João Silva", "joao@example.com");

        MvcResult createResult = mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isCreated())
                .andReturn();

        String responseContent = createResult.getResponse().getContentAsString();
        Long userId = objectMapper.readTree(responseContent).get("id").asLong();

        UpdateUserRequest updateRequest = new UpdateUserRequest("João Silva Updated", "joao.updated@example.com");

        mockMvc.perform(put("/api/users/{id}", userId)
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(updateRequest)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(userId))
                .andExpect(jsonPath("$.name").value("João Silva Updated"))
                .andExpect(jsonPath("$.email").value("joao.updated@example.com"));
    }

    @Test
    void shouldDeleteUserSuccessfully() throws Exception {
        CreateUserRequest createRequest = new CreateUserRequest("João Silva", "joao@example.com");

        MvcResult createResult = mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isCreated())
                .andReturn();

        String responseContent = createResult.getResponse().getContentAsString();
        Long userId = objectMapper.readTree(responseContent).get("id").asLong();

        mockMvc.perform(delete("/api/users/{id}", userId))
                .andExpect(status().isNoContent());

        mockMvc.perform(get("/api/users/{id}", userId))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.message").value("Usuário não encontrado com ID: " + userId));
    }

    @Test
    void shouldGetAllUsersSuccessfully() throws Exception {
        CreateUserRequest user1 = new CreateUserRequest("João Silva", "joao@example.com");
        CreateUserRequest user2 = new CreateUserRequest("Maria Santos", "maria@example.com");

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(user1)))
                .andExpect(status().isCreated());

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(user2)))
                .andExpect(status().isCreated());

        mockMvc.perform(get("/api/users"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").isArray())
                .andExpect(jsonPath("$.length()").value(2))
                .andExpect(jsonPath("$[0].name").value("João Silva"))
                .andExpect(jsonPath("$[1].name").value("Maria Santos"));
    }

    @Test
    void shouldGetUserByEmailSuccessfully() throws Exception {
        CreateUserRequest createRequest = new CreateUserRequest("João Silva", "joao@example.com");

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(createRequest)))
                .andExpect(status().isCreated());

        mockMvc.perform(get("/api/users/email/{email}", "joao@example.com"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.name").value("João Silva"))
                .andExpect(jsonPath("$.email").value("joao@example.com"));
    }

    @Test
    void shouldReturnBadRequestForInvalidUserData() throws Exception {
        CreateUserRequest invalidRequest = new CreateUserRequest("", "invalid-email");

        mockMvc.perform(post("/api/users")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(invalidRequest)))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.errors.name").exists())
                .andExpect(jsonPath("$.errors.email").exists());
    }

    @Test
    void shouldReturnNotFoundForNonExistentUser() throws Exception {
        mockMvc.perform(get("/api/users/{id}", 999L))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.message").value("Usuário não encontrado com ID: 999"));
    }

    @Test
    void shouldReturnNotFoundForNonExistentEmail() throws Exception {
        mockMvc.perform(get("/api/users/email/{email}", "nonexistent@example.com"))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.message").value("Usuário não encontrado com email: nonexistent@example.com"));
    }
}

File: src/test/kotlin/com/gisiona/demodevinia/integration/UserIntegrationTest.kt
package com.gisiona.demodevinia.integration

import com.gisiona.demodevinia.infrastructure.web.dto.CreateUserRequest
import com.gisiona.demodevinia.infrastructure.web.dto.UpdateUserRequest
import com.gisiona.demodevinia.infrastructure.web.dto.UserResponse
import org.junit.jupiter.api.*
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpEntity
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
@ActiveProfiles("test")
@Transactional
class UserIntegrationTest {

    @Autowired
    private lateinit var testRestTemplate: TestRestTemplate

    @Test
    @Order(1)
    fun `should perform complete user CRUD operations`() {
        val createRequest = CreateUserRequest("John Doe", "john@example.com")

        val createResponse = testRestTemplate.postForEntity(
            "/api/users",
            createRequest,
            UserResponse::class.java
        )

        Assertions.assertEquals(HttpStatus.CREATED, createResponse.statusCode)
        Assertions.assertNotNull(createResponse.body)
        Assertions.assertEquals("John Doe", createResponse.body!!.name)
        Assertions.assertEquals("john@example.com", createResponse.body!!.email)

        val userId = createResponse.body!!.id

        val getResponse = testRestTemplate.getForEntity(
            "/api/users/$userId",
            UserResponse::class.java
        )
        Assertions.assertEquals(HttpStatus.OK, getResponse.statusCode)
        Assertions.assertEquals("John Doe", getResponse.body!!.name)

        val getAllResponse = testRestTemplate.getForEntity(
            "/api/users",
            Array<UserResponse>::class.java
        )
        Assertions.assertEquals(HttpStatus.OK, getAllResponse.statusCode)
        Assertions.assertTrue(getAllResponse.body!!.isNotEmpty())

        val updateRequest = UpdateUserRequest("John Smith", null)
        val updateResponse = testRestTemplate.exchange(
            "/api/users/$userId",
            HttpMethod.PUT,
            HttpEntity(updateRequest),
            UserResponse::class.java
        )
        Assertions.assertEquals(HttpStatus.OK, updateResponse.statusCode)
        Assertions.assertEquals("John Smith", updateResponse.body!!.name)

        val deleteResponse = testRestTemplate.exchange(
            "/api/users/$userId",
            HttpMethod.DELETE,
            null,
            Void::class.java
        )
        Assertions.assertEquals(HttpStatus.NO_CONTENT, deleteResponse.statusCode)

        val getDeletedResponse = testRestTemplate.getForEntity(
            "/api/users/$userId",
            UserResponse::class.java
        )
        Assertions.assertEquals(HttpStatus.NOT_FOUND, getDeletedResponse.statusCode)
    }

    @Test
    @Order(2)
    fun `should prevent duplicate email creation`() {
        val createRequest = CreateUserRequest("Jane Doe", "jane@example.com")

        val firstResponse = testRestTemplate.postForEntity(
            "/api/users",
            createRequest,
            UserResponse::class.java
        )
        Assertions.assertEquals(HttpStatus.CREATED, firstResponse.statusCode)

        val secondResponse = testRestTemplate.postForEntity(
            "/api/users",
            createRequest,
            String::class.java
        )
        Assertions.assertEquals(HttpStatus.CONFLICT, secondResponse.statusCode)
    }
}
